Add @cruise-automation/hooks package (#116)
cruise-automation/webviz: packages/@cruise-automation/hooks/README.md

# @cruise-automation/hooks

A set of resusable React Hooks.

## Install

```bash
npm install --save @cruise-automation/hooks
```

## Hooks

### `useAbortable`

A React Hook to load async, disposable resources and fire the cleanup callback when the component unmountts. If the component unmounts before the async operation completes, the resource will still be cleaned up once it finishes loading and an abort signal will be issued to the async load operation.

```js
// types
 useAbortable<T>(
  defaultValue: T,  // default value, will be set to the actual value if action is performed successfully
  action: (AbortController) => Promise<T>, // async action to perform
  cleanup: (?T) => void, // clean up work if thtere is any
  args: any // dependencies
): [T, () => void] // result value and abort function
```

```js
// sample usage
import { useAbortable } from "@cruise-automation/hooks";

function Example(props) {

>   const [data, setDatta] = useState(null);

Reviewer: ```suggestion
  const [data, setData] = useState(null);
```

>   const dateName = 'dateName';

Reviewer: ```suggestion
  const dataName = 'foo';
```

  // fetch data from remote when the component is mounted
  const [remoteData, abortFn] = useAbortable(
    null,
    async (abortController) =>  {

>       if(dataName) {

Reviewer: ```suggestion
      if (dataName) {
```

        fetchDataFromRemote(props, dataName)
      }
    }
    (val) => {},
    [dataName]
  );

  // abort is usually called when the component unmounts, but it can also be called manually
  function abortManually() {
    abortFn();
  }

  return (
    <div>

>       <button onClick={abortManually}> Abort Manually Now</button>

Reviewer: ```suggestion
      <button onClick={abortManually}>Stop Loading</button>
```

    </div>
  );
}
```

### `useCleanup`

> A small React Hook to fire the cleanup callback when the component unmounts.

Reviewer: ```suggestion
A small React Hook to fire the cleanup callback when the component unmounts. Equivalent to `useEffect(() => () => { teardown(); }, [])`.
```

```js
// types
 useCleanup(teardown: () => void): void
```

```js
// sample usage
function Example() {
  const [audioContext] = React.useState(() => new window.AudioContext));
  // automatically close audioContext when the component unmounts
  useCleanup(() => audioContext.close());
  return null;
}
```

### `useEventListener`

> A handy React Hook for automatically adding and removing event listeners.

Reviewer: ```suggestion
A hook for conditionally adding and removing event listeners on DOM elements.
```

```js
// types
useEventListener<T>(
  target: Element, // event target, e.g. window
  type: string, // event type
  enable: boolean,
  handler: (any) => void,
  dependencies: any[],
): void
```

```js
// sample usage
function Example() {
  const [isDragging, setIsDragging] = useState(false);
  useEventListener(
    window,
    "mousemove",

>     isDragging, // enable during mouse dragging

Reviewer: ```suggestion
    isDragging, // add this listener only during mouse dragging
```

    (event) => {
      /* do something here... */
    },
    []
  );
  return <div onMouseDown={() => setIsDragging(true)}>demo</div>;
}
```

### `useAnimationFrame`

> A React Hook that accepts a callback function which will be called before each animation frame.

Reviewer: ```suggestion
A React Hook that accepts a callback function which will be called repeatedly, synchronized with the browser's repaints via [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame).
```

```js
// types
useAnimationFrame(
  callback: (timestamp: number) => void,

>   disable: boolean,

Reviewer: it's a bit weird to use `enable` in useEventListener but `disable` in useAnimationFrame.

Author: For `useEventListener`, the consumer would normally do some sort of `isDragging` state  => `enable` listener mapping, and there is need to think about defaults, so keeping `enable` prop name makes sense.  

For `useAnimationFrame`, if the user doesn't pass in `disable` prop, we'll just enable `useAnimationFrame` by default which is what's expected if people are using this hook. 

So overall, yeah, it's weird, but maybe not enough of a reason to keep them the same?

  dependencies: any[],
): void
```

```js

> // sample useage: a count state that gets updated in every animation frame

Reviewer: ```suggestion
// sample usage: a count state that gets updated in every animation frame
```

function Example() {
  const [count, setCount] = React.useState(0);
  useAnimationFrame(

>     () => {

Reviewer: ```suggestion
    (timestamp) => {
```

# @cruise-automation/hooks

A set of resusable React Hooks.

## Usage

```bash
npm install --save @cruise-automation/hooks
```

### useAbortable

```js
import { useAbortable, useConstant, useEventListener, useRequestAnimationFrame } from "@cruise-automation/hooks";

// Abort an fetchData action

> const [remoteData] = useAbortable(

Reviewer: Second return value (abort function) should not be ignored.

# @cruise-automation/hooks

A set of resusable React Hooks.

## Usage

```bash
npm install --save @cruise-automation/hooks
```

```js
import { useAbortable, useCleanup, useEventListener, useAnimationFrame } from "@cruise-automation/hooks";

> //*************** useAbortable **********************

Reviewer: why not use markdown headings `###` for these?

// Abort an fetchData action
const [remoteData, abortFn] = useAbortable(
  [], // default value
  async (abortController) => fetchDataFromRemote(props, "dataName"), // action to perform
  (val) => {}, // clean up work if there is any
  ["dataName"] // remount when dependencies change
);

> // call abortFn somewhere else, e.g. before component unmounts

Reviewer: Doesn't useAbortable already call it when the component is unmounting?

abortFn();

//*************** useCleanup **********************
// Call the cleanup function before the component unmounts

> const [audioContent] = React.useState(() => new window.AudioContext));

Reviewer: ```suggestion
const [audioContext] = React.useState(() => new window.AudioContext));
```
